### PR TITLE
SIG Scalability Charter

### PR DESCRIPTION
**Pull request by wojtek-t**
@kubernetes/sig-scalability-api-reviews 
@countspongebob @shyamjvs @wasylkowski-a @gmarek 

cc-ing everyone who was on the initial version:
/cc @thockin @smarterclayton @spiffxp @jdumars @bgrant0607 @timothysc

### CODE REVIEW

File: sig-scalability/charter.md [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+### Technical Lead
+- Number: 2-3
+- Establish new subprojects and retire existing ones
+- Resolve cross-subprojects technical issues and decisions and escalations from
+  subprojects
+- Decision making must be by consensus.
+- An initial set of technical leads was set to long-standing group of SIG leads:
+  Wojciech Tyczynski and Bob Wise
+- Technical leads must have demonstrated deep understanding of the whole system
+  that is sufficient to access impact of different changes on Kubernetes scalability.
```
Comment (wasylkowski-a) [inline sig-scalability/charter.md]: s/access/assess

File: sig-scalability/charter.md [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+The following roles are required for the SIG to properly function.
+In the event that any role is unfilled, the SIG will make a best effort
+to fill it and any decisions reliant on a missing role will be postponed
+until the role is filled.
+
+### Chair
+- Number: 2-3
+- Run operations and processes governing the SIG
+- A majority of chairs cannot be from a single company.
+- An initial set of chairs was established at the time the SIG was founded.
```
Comment (wasylkowski-a) [inline sig-scalability/charter.md]: Who are the initial chairs?

File: sig-scalability/charter.md (hunk 1 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: done

File: sig-scalability/charter.md (hunk 2 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Bob Wise and myself

Comment (wasylkowski-a) [reply sig-scalability/charter.md]: Can you put it into the charter given that you have put TLs' names?

**Comment by wojtek-t** [comment]
@thockin @smarterclayton @bgrant0607 @spiffxp - can you please take a look?
Is there any official procedure to have the charter approved?

**Comment by jdumars** [comment]
This is in the Steering Committee backlog under: https://github.com/kubernetes/steering/issues/31

File: sig-scalability/charter.md [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+performance improvements (that don’t fall into the charter of another individual
+SIG) as well as provide consultations about any scalability and performance
+related aspects of Kubernetes.
+
+## What can we do/require from other SIGs
+Scalability and performance are horizontal aspects of the system - changes in a
+single place of Kubernetes may affect the whole system. As a result, to
+effectively ensure Kubernetes scales, we need a special cross-SIG privileges.
+
+- We can rollback any merged PR if it has been identified as a cause of an SLO
```
Comment (philips) [inline sig-scalability/charter.md]: can you link to the SLOs? Where are they? Who defines them? Is the API machinery one below the only example?

File: sig-scalability/charter.md [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+<tr>
+  <td>Scalability testing frameworks</td>
+  <td>Designing and creating frameworks to make scalability and performance testing
+  of Kubernetes easy and available for all contributors. Different frameworks may
+  help in different aspect of scalability testing enabling making conscious tradeoffs,
+  e.g. cost of accuracy or real life vs more generalized benchmarking scenarios. </td>
+  <td>Cluster loader</td>
+</tr>
+<tr>
+  <td>Scalability and performanc tests</td>
```
Comment (philips) [inline sig-scalability/charter.md]: perfomanc spelling error

File: sig-scalability/charter.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  regression. The offending PR should only be merged again after proving to pass
+  tests at scale.
+- We can pause the merge queue in case of a regression observed until a particular
+  PR has been identified as cause of the regression and regression has been
+  mitigated. The “Rules of engagement” of pausing merge-queue and rationale for
+  necessity of its introduce are explained in a separate doc. <br/>
+  TODO(wojtek-t, shyamjvs): Write it down and link here.
+- We require significant changes (in terms of impact, such as: update of etcd,
+  update of Go version, major architectural changes, etc.) may only be merged:
+  - with an explicit approval from a SIG-scalability approver and
```
Comment (philips) [inline sig-scalability/charter.md]: where are the approvers documented?

File: sig-scalability/charter.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+performance. However, all issues that fully fall under a single SIG are implicitly
+delegated to that SIG.
+
+SIG scalability subprojects are as follows.
+
+<table>
+<tr>
+  <td>Subproject</td>
+  <td>Description</td>
+  <td>Example Artifacts</td>
```
Comment (philips) [inline sig-scalability/charter.md]: Can this table have a column for a link to the OWNERS file? It requires a bit of digging for each of these.

Comment (wojtek-t) [reply sig-scalability/charter.md]: Done (modulo one line, which I would like to figure out independently in parallel - hopefully that doesn't block charter).

File: sig-scalability/charter.md (hunk 2 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: done

File: sig-scalability/charter.md (hunk 5 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Added subsection in the roles section.

File: sig-scalability/charter.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+related aspects of Kubernetes.
+
+## What can we do/require from other SIGs
+Scalability and performance are horizontal aspects of the system - changes in a
+single place of Kubernetes may affect the whole system. As a result, to
+effectively ensure Kubernetes scales, we need a special cross-SIG privileges.
+
+- We can rollback any merged PR if it has been identified as a cause of any
+  [performance/scalability SLOs] regression. The offending PR should only be
+  merged again after proving to pass  tests at scale.
```
Comment (philips) [inline sig-scalability/charter.md]: Is that an automated process? e.g. as a PR owner do I know if my code has violated SLOs?

Comment (wojtek-t) [reply sig-scalability/charter.md]: Partially. So we already have some presubmits that are checking that on smaller scale.
And for larger scale (i.e. 5k nodes) it's impossible to have presubmits because they are way too long and way too expensive.
So that is being checked on postsubmits and as of now monitores by sig-scalability. We are working on introducing more automation here, but that's not yet done.

**Comment by philips** [comment]
Thank you for your patience as we work to your SIG charter merged.

Through the process of reviewing the [first 11 charters][charters] the Steering Committee (SC) found a lot of copy/paste language caused by our poorly designed initial charter template. To fix this there is a new [SIG charter template][template] which focuses on the main things we want to see in charters: scope and responsibilities.

Please consider updating to this charter template and putting focus into the scope and responsibilities. SC members will focus more on a deep review of scope more than anything else and using this template will help with that focus.

[charters]: https://docs.google.com/spreadsheets/d/1i_8oM1KlmXo90zIIgNTdOg-M6xesTggFLnTkgmscMlI/edit#gid=0
[template]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-charter-template.md

**Comment by wojtek-t** [comment]
@philips - Thanks Brandon. I'm currently on paternity leave and will be back around in ~4-5 weeks from now. If this can't be merged in the current state, I will get back to it once I'm back.
Hope that's fine for you too.

**Comment by spiffxp** [comment]
@wojtek-t I'm on file as one of the reviewers for this, I can trying poking @countspongebob if you would prefer

**Comment by wojtek-t** [comment]
@philips @timothysc @spiffxp - I refactored the previous version to better match the template. Hope that it's close to what you expect.

PTAL

@countspongebob @shyamjvs - FYI

File: sig-scalability/charter.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+This charter adheres to the conventions described in the [Kubernetes Charter README]
+and uses the Roles and Organization Management outlined in [sig-governance].
+
+[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md
+[Kubernetes Charter README]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md
+
+## Scope
+
+SIG Scalability primary responsibilities are to define scalability goals for
+Kubernetes, define, test and measure performance and scalability related
```
Comment (timothysc) [inline sig-scalability/charter.md]: s/,define,/

File: sig-scalability/charter.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+Service Level Indicators (SLIs) and ensure that every Kubernetes release meets
+Service Level Objectives (SLOs) build on top of the above SLIs.
+
+We also coordinate and contribute to general system-wide scalability and
+performance improvements (that don't fall into the charter of another individual
+SIG) as well as provide consultations about any scalability and performance
+related aspects of Kubernetes.
```
Comment (timothysc) [inline sig-scalability/charter.md]: There is also large refactoring/ architectural restructuring work that is done here too, please call that out.

File: sig-scalability/charter.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+- Scalability and performance testing frameworks. Examples include:
+  - [Cluster loader](https://github.com/kubernetes/perf-tests/tree/master/clusterloader2)
+  - [Kubemark](https://github.com/kubernetes/kubernetes/tree/master/cmd/kubemark)
+- Scalability and performance tests:
+  - [Tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scalability/)
+
+#### Cross-cutting and Externally Facing Processes
+
+- Defining what does it mean that “Kubernetes scales”. This includes defining
```
Comment (timothysc) [inline sig-scalability/charter.md]: I'd nix the 1st part and just outline the SLIs and SLOs.

File: sig-scalability/charter.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+- Ensuring that each official Kubernetes release satisfies all scalability and
+performance related requirements, as stated in "Kubernetes scalability" definition.
+- Establishing and documenting best practises on how to design and/or implement
+Kubernetes features in scalable and performant way. Educating contributors and
+consulting individual designs/implementations to ensure that those are widely used.
+Example artifacts:
+  - [Scalability governance](https://github.com/kubernetes/community/blob/master/sig-scalability/governance)
+
+
+TODO: Figure out how to phrase finding bottlenecks, coordinating improvements
```
Comment (timothysc) [inline sig-scalability/charter.md]: I mentioned this above.

File: sig-scalability/charter.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+and architectural changes, etc.
+
+### Out of scope
+
+- Improving performance/scalability of features falling into charters of
```
Comment (timothysc) [inline sig-scalability/charter.md]: Really? b/c we've really refactored the heck out of some features.

File: sig-scalability/charter.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+- We can require a SIG to introduce a regression-catching benchmark test for a
+  scalability-critical functionality.
+
+For the record, by regression above we mean a regression identified by the set 
+of release-blocking scalability/performance tests (as defined by
+sig-release-master-blocking group of test suites).
+
+[performance/scalability SLOs]: https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md
+
+### SIG Scalability approvers
```
Comment (timothysc) [inline sig-scalability/charter.md]: this shouldn't be here.  
this should be in the sigs.yaml

File: sig-scalability/charter.md [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+## Roles and Organization Management
+
+This sig follows adheres to the Roles and Organization Management outlined in
+[sig-governance] and opts-in to updates and modifications to [sig-governance].
+
+[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md
+
+### Subproject Creation
+
+SIG Auth delegates subproject approval to Technical Leads. See [Subproject creation - Option 1].
```
Comment (timothysc) [inline sig-scalability/charter.md]: Auth?

File: sig-scalability/charter.md (hunk 8 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Why? We actually define SLIs/SLOs (with cooperation with other sigs).

File: sig-scalability/charter.md (hunk 10 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: I think that the fact that this is the definition of "Kubernetes scalability" is relatively important.
I rephrased it a bit - ptal.

File: sig-scalability/charter.md (hunk 9 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Done - ptal.

File: sig-scalability/charter.md (hunk 11 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Done.

File: sig-scalability/charter.md (hunk 12 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Agree - though we don't have enough capacity to do that (it shouldn't be out missing).
Though in some cases we may help with that...

File: sig-scalability/charter.md (hunk 14 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: copy-paste bug :)

File: sig-scalability/charter.md (hunk 8 above)
Comment (timothysc) [reply sig-scalability/charter.md]: Gramatically the sentence is weird.  

perhaps 
`SIG Scalability primary responsibilities are to define scalability goals for
Kubernetes.  This involves defining ...`

**Comment by spiffxp** [comment]
/hold
for comment

File: sig-scalability/charter.md (hunk 8 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Thanks for suggestion - done.

File: sig-scalability/charter.md (hunk 13 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: So I thought a bit more about that, and:
- longer term it may make sense to move to sigs.yaml
- currently, it's not supported there (it's somewhat sig-specific role) and i don't want to add that now.

Comment (timothysc) [reply sig-scalability/charter.md]: Yes it is, please see other sigs and remove.

Comment (wojtek-t) [reply sig-scalability/charter.md]: Can you please point me to one example? I claim that anything that exists is not what we want (i.e. this list may contain people that aren't even scalability sig members, but they have wide system knowledge and can access scalability impact of changes).

Comment (timothysc) [reply sig-scalability/charter.md]: We list the "chairs" and "leads" in the main sigs.yaml, approvers is a concept in the OWNERs which is subject to change over time.  Eventually, folks want to get to a canonical location for the OWNERs, but the charter is not the appropriate location for what is essentially "OWNERs data", and has been strip'd out from other charters.

File: sig-scalability/charter.md [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+This charter adheres to the conventions described in the [Kubernetes Charter README]
+and uses the Roles and Organization Management outlined in [sig-governance].
+
+[sig-governance]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-governance.md
+[Kubernetes Charter README]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/README.md
+
+## Scope
+
+SIG Scalability primary responsibilities are to define scalability goals for
+Kubernetes. This involves defining, testing and measuring performance and
```
[the remark this answers is not in the document]
Comment (spiffxp) [reply sig-scalability/charter.md]: It may be worth considering taking an approach similar to SIG Instrumentation, another horizontal SIG with some areas of discrete ownership:
- Own performance and scalability best practices for the project as a whole
- Provide reviews, and guidance to other SIGs
- Own the definition, implementation, and enforcement of scalability SLIs and SLOs for the artifacts produced as part of a Kubernetes release

File: sig-scalability/charter.md [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+### In Scope
+
+#### Code, Binaries and Services:
+
+- Scalability and performance testing frameworks. Examples include:
+  - [Cluster loader](https://github.com/kubernetes/perf-tests/tree/master/clusterloader2)
+  - [Kubemark](https://github.com/kubernetes/kubernetes/tree/master/cmd/kubemark)
+- Scalability and performance tests:
+  - [Tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scalability/)
+
```
Comment (spiffxp) [inline sig-scalability/charter.md]: To be clear, there are no areas of the kubernetes/kubernetes codebase that fall under the responsibility of SIG scalability?

File: sig-scalability/charter.md [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+### In Scope
+
+#### Code, Binaries and Services:
+
+- Scalability and performance testing frameworks. Examples include:
+  - [Cluster loader](https://github.com/kubernetes/perf-tests/tree/master/clusterloader2)
+  - [Kubemark](https://github.com/kubernetes/kubernetes/tree/master/cmd/kubemark)
+- Scalability and performance tests:
+  - [Tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/scalability/)
```
Comment (spiffxp) [inline sig-scalability/charter.md]: You also own the jobs that run these tests: https://github.com/kubernetes/test-infra/tree/master/config/jobs/kubernetes/sig-scalability

File: sig-scalability/charter.md [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+
+## What can we do/require from other SIGs
+Scalability and performance are horizontal aspects of the system - changes in a
+single place of Kubernetes may affect the whole system. As a result, to
+effectively ensure Kubernetes scales, we need a special cross-SIG privileges.
+
+- We can rollback any merged PR if it has been identified as a cause of any
+  [performance/scalability SLOs] regression. The offending PR should only be
+  merged again after proving to pass  tests at scale.
+- We can pause the merge queue in case of a regression observed until a particular
```
Comment (spiffxp) [inline sig-scalability/charter.md]: The concept of a merge queue is going away.  I would rephrase this as "in the event of a performance regression of the kubernetes release artifacts, we can block all PRs from being merged into the relevant repos until the offending PR is identified and reverted" and link to the more formal doc that spells out exactly: which branches you can block, which repos you can block, how/whether this policy changes over the release lifecycle, etc.

File: sig-scalability/charter.md (hunk 7 above)
Comment (spiffxp) [reply sig-scalability/charter.md]: We don't necessarily have to do this within this charter, but I'm going to suggest that the release team require regular participation from a member of SIG Scalability until we have such automation in place.  Chasing down scalability related failures is too much an additional burden on the CI Signal role, and we often have little visibility into whether or how failures are being worked.

As an example both https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-scale-performance and https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-scale-correctness have been failing for over two weeks.  This is being "explained away" as a quota issue, with code continuing to merge, rather than being enforced as a hard stop until the cause is identified.

File: sig-scalability/charter.md [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+  PR has been identified as cause of the regression and regression has been
+  mitigated. The “Rules of engagement” of pausing merge-queue and rationale for
+  necessity of its introduce are explained in a separate doc. <br/>
+  TODO(wojtek-t, shyamjvs): Write it down and link here.
+- We require significant changes (in terms of impact, such as: update of etcd,
+  update of Go version, major architectural changes, etc.) may only be merged:
+  - with an explicit approval from a [SIG-scalability approver](#sig-scalability-approvers)
```
Comment (spiffxp) [inline sig-scalability/charter.md]: tech lead

File: sig-scalability/charter.md [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+    and
+  - after having passed performance testing on biggest supported clusters (unless
```
Comment (spiffxp) [inline sig-scalability/charter.md]: In a future where there are large cluster jobs being run on multiple cloud providers, will you require that they all pass?

How does someone run these jobs as a presubmit?

File: sig-scalability/charter.md [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+    found unnecessary by scalability approver)
+- We can block a feature from transitioning to Beta status if (when turned on) it
+  causes a significant degradation of overall Kubernetes scalability/performance.
+  (Ideally it would be “SLI degradation of more than X%” or “breaking SLO”, but
+  initially it may also be SIG-scalability decision based on public test results).
+- We can block a feature from transitioning to GA status if it cannot be used at
```
Comment (spiffxp) [inline sig-scalability/charter.md]: How is this different than the requirements spelled out for beta status? I'd like to see this phrased in terms of SLI/SLO

File: sig-scalability/charter.md [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+  mitigated. The “Rules of engagement” of pausing merge-queue and rationale for
+  necessity of its introduce are explained in a separate doc. <br/>
+  TODO(wojtek-t, shyamjvs): Write it down and link here.
+- We require significant changes (in terms of impact, such as: update of etcd,
+  update of Go version, major architectural changes, etc.) may only be merged:
+  - with an explicit approval from a [SIG-scalability approver](#sig-scalability-approvers)
+    and
+  - after having passed performance testing on biggest supported clusters (unless
+    found unnecessary by scalability approver)
+- We can block a feature from transitioning to Beta status if (when turned on) it
```
Comment (spiffxp) [inline sig-scalability/charter.md]: I _really like_ this idea.  But, you might want to consider rephrasing it such that SIG Architecture actually owns this enforcement, and you are in charge of defining the scalability SLI's/SLO's that they use.  Ultimately I think it's going to be less confusing for feature owners to look to a single SIG for go/no-go, and use a checklist of requirements you've helped write.

File: sig-scalability/charter.md (hunk 13 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: I gave up for now and switched to tech leads with a TODO to create a dedicated group for it.

File: sig-scalability/charter.md (hunk 7 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: It's definitely not your role (as sig release) to track the failures - I agree.

But again, whether we should be on sig release meetings or not, is an implementation detail. We are responsible for those tests and it's possible that this means that we need to come to sig release.
But that's implementation detail of the mission, not part of the charter.

File: sig-scalability/charter.md [hunk 23]
```diff
@@ -0,0 +1,104 @@
...
+
+- We can rollback any merged PR if it has been identified as a cause of any
+  [performance/scalability SLOs] regression. The offending PR should only be
+  merged again after proving to pass  tests at scale.
+- In the even of a performance regression, we can block all PRs from being
+  merged into the relevant repos until the cause of the regression is
+	identified and mitigated.
+  The “Rules of engagement” of pausing merge-queue and rationale for
+  necessity of its introduce are explained in a separate doc. <br/>
+  TODO(wojtek-t, shyamjvs): Write it down and link here.
```
Comment (timothysc) [inline sig-scalability/charter.md]: ?

File: sig-scalability/charter.md [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+  - to Beta status, if (when turned on) it causes violation of already existing
+    performance/scalability SLOs;
+  - to GA status, when it can be used scale. That means:
+    - in rare cases, introducing a new SLI and SLO and ensuring it is met at scale
+    - in most of cases, extending scalability tests to use it and ensuring that
+      existing SLOs are still met
+- We can require a SIG to introduce a regression-catching benchmark test for a
+  scalability-critical functionality.
+
+For the record, by regression above we mean a regression identified by the set
```
Comment (timothysc) [inline sig-scalability/charter.md]: might just want to define taxonomy at the beginning or end, this seems weirdly placed.  Could just remove.

File: sig-scalability/charter.md (hunk 23 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Added.

**Comment by wojtek-t** [comment]
PTAL

File: sig-scalability/charter.md [hunk 25]
```diff
@@ -0,0 +1,100 @@
...
+  merged into the relevant repos until the cause of the regression is
+	identified and mitigated.
+  The “Rules of engagement” of pausing merge-queue and rationale for
+  necessity of its introduce are explained in [a separate doc](./block_merges.md).
+- We require significant changes (in terms of impact, such as: update of etcd,
+  update of Go version, major architectural changes, etc.) may only be merged:
+  - with an explicit approval from a SIG-scalability tech lead and
+  - after having passed performance testing on biggest supported clusters (unless
+    found unnecessary by approver)
+  TODO(wojtek-t): Create a larger group (whose members doesn't have to necessary
```
Comment (timothysc) [inline sig-scalability/charter.md]: TODO - still exists.

Comment (wojtek-t) [reply sig-scalability/charter.md]: I don't want to resolve it now - this is a bigger thing and for now I think it's fine to start with tech leads.

Comment (timothysc) [reply sig-scalability/charter.md]: I'd just remove that sentence then, and PR the charter at a later time when TODO is less ambiguous.

File: sig-scalability/block_merges.md [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+- Identify the PR which caused the regression:
+  - this can be done by reading code changes, bisecting, debugging based on
+    metrics and/or logs, etc.
+  - we say a PR is identified as the cause when we are reasonably confident
+    that it indeed caused a regression, even if the mechanism is not 100%
+    understood to minimize the time when merges are blocked
+- Mitigate the regression. This may mean e.g.:
+  - reverting the PR
+  - switching a feature off (preferably by default, as last resort only in tests)
+  - ...
```
Comment (philips) [inline sig-scalability/block_merges.md]: Please remove this `...`. It doesn't add anything to the list.

Comment (wojtek-t) [reply sig-scalability/block_merges.md]: fair enough - done.

File: sig-scalability/charter.md (hunk 25 above)
Comment (wojtek-t) [reply sig-scalability/charter.md]: Make sense - done.

File: sig-scalability/charter.md [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+## What can we do/require from other SIGs
+Scalability and performance are horizontal aspects of the system - changes in a
+single place of Kubernetes may affect the whole system. As a result, to
+effectively ensure Kubernetes scales, we need a special cross-SIG privileges.
+
+- We can rollback any merged PR if it has been identified as a cause of any
+  [performance/scalability SLOs] regression (identified by the set of release
+  blocking scalability/performance tests). The offending PR should only be
+  merged again after proving to pass  tests at scale.
+- In the even of a performance regression, we can block all PRs from being
```
Comment (spiffxp) [inline sig-scalability/charter.md]: nit: event

File: sig-scalability/charter.md [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+  merged into the relevant repos until the cause of the regression is
+	identified and mitigated.
```
Comment (spiffxp) [inline sig-scalability/charter.md]: nit: whitespace

File: sig-scalability/charter.md [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+  The “Rules of engagement” of pausing merge-queue and rationale for
+  necessity of its introduce are explained in [a separate doc](./block_merges.md).
+- We require significant changes (in terms of impact, such as: update of etcd,
+  update of Go version, major architectural changes, etc.) may only be merged:
+  - with an explicit approval from a SIG-scalability tech lead and
```
Comment (spiffxp) [inline sig-scalability/charter.md]: I don't see your tech leads in sigs.yaml

File: sig-scalability/block_merges.md [hunk 30]
```diff
@@ -0,0 +1,51 @@
+# Blocking PR merges in the event of regression.
+
+As mentioned in the charter, SIG scalability has a right to block all PRs
+from merging into the relevant repos. This document describes the underlying
+"Rules of engagement" of this process and the rationale why this is needed.
+
+### Rules of engagement.
+The rules of engagement for blocking merges are as following:
+
+- Observe as scalability regression on one of release-blocking test suites.
```
Comment (spiffxp) [inline sig-scalability/block_merges.md]: Which suites specifically?

How is a regression observed?  The suites have been continually failing, so it doesn't seem like it's a green-to-red transition.

File: sig-scalability/block_merges.md [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+  - this can be done by reading code changes, bisecting, debugging based on
+    metrics and/or logs, etc.
+  - we say a PR is identified as the cause when we are reasonably confident
+    that it indeed caused a regression, even if the mechanism is not 100%
+    understood to minimize the time when merges are blocked
+- Mitigate the regression. This may mean e.g.:
+  - reverting the PR
+  - switching a feature off (preferably by default, as last resort only in tests)
+  - fixing the problem (if it's easy and quick to fix)
+- Unblock PR merged.
```
Comment (spiffxp) [inline sig-scalability/block_merges.md]: nit: "Unblock merges of all PRs to the relevant repos"

File: sig-scalability/block_merges.md [hunk 32]
```diff
@@ -0,0 +1,51 @@
...
+
+As mentioned in the charter, SIG scalability has a right to block all PRs
+from merging into the relevant repos. This document describes the underlying
+"Rules of engagement" of this process and the rationale why this is needed.
+
+### Rules of engagement.
+The rules of engagement for blocking merges are as following:
+
+- Observe as scalability regression on one of release-blocking test suites.
+- Block merges of all PRs.
```
Comment (spiffxp) [inline sig-scalability/block_merges.md]: Block merges of all PRs to the relevant repos, declaring which repos those are and why

File: sig-scalability/block_merges.md [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+  - we say a PR is identified as the cause when we are reasonably confident
+    that it indeed caused a regression, even if the mechanism is not 100%
+    understood to minimize the time when merges are blocked
+- Mitigate the regression. This may mean e.g.:
+  - reverting the PR
+  - switching a feature off (preferably by default, as last resort only in tests)
+  - fixing the problem (if it's easy and quick to fix)
+- Unblock PR merged.
+
+The exact technical mechanisms for it are out of scope for this document.
```
Comment (spiffxp) [inline sig-scalability/block_merges.md]: If not here then where are these mechanisms documented?

File: sig-scalability/block_merges.md [hunk 34]
```diff
@@ -0,0 +1,51 @@
...
+if we want kubernetes to maintain scalability SLOs. The reasoning is:
+- reliably testing for regressions takes a lot of time:
+  - key scalability e2e tests take too long to execute to be a prerequisite
+    for merging all PRs, this is an inherent characteristic of testing at scale,
+  - end-to-end tests are flaky (even when not at scale) requiring retries,
+- we need to prevent regression pile-ups:
+  - once a regression is merged, and no other action is taken, it is only
+    a matter of time until another regression is merged on top of it,
+  - debugging the cause of two simultaneous (piled-up) regressions is 
+    exponentially harder, see issue 53255 which links to past experience
```
Comment (spiffxp) [inline sig-scalability/block_merges.md]: nit: link to https://github.com/kubernetes/kubernetes/issues/53255

File: sig-scalability/block_merges.md [hunk 35]
```diff
@@ -0,0 +1,51 @@
...
+- we need to keep flakiness of merge-blocking jobs very low:
```
Comment (spiffxp) [inline sig-scalability/block_merges.md]: what is the threshold? I suspect we are not meeting this today, so I would like to quantify how we're approaching this delta from actual to intended

**Comment by spiffxp** [comment]
@timothysc Next time please give a last call ping prior to merging
